server/channels/store/sqlstore/reaction_store.go

```diff
@@ -205,24 +205,70 @@ func (s *SqlReactionStore) DeleteAllWithEmojiName(emojiName string) error {
 	return nil
 }
 
-// DeleteOrphanedRows removes entries from Reactions when a corresponding post no longer exists.
-func (s *SqlReactionStore) DeleteOrphanedRows(limit int) (deleted int64, err error) {
-	// We need the extra level of nesting to deal with MySQL's locking
-	const query = `
-	DELETE FROM Reactions WHERE PostId IN (
-		SELECT * FROM (
-			SELECT PostId FROM Reactions
-			LEFT JOIN Posts ON Reactions.PostId = Posts.Id
-			WHERE Posts.Id IS NULL
-			LIMIT ?
-		) AS A
-	)`
-	result, err := s.GetMasterX().Exec(query, limit)
+func (s SqlReactionStore) PermanentDeleteByUser(userId string) error {
+	now := model.GetMillis()
+	postIds := []string{}
+
+	err := s.GetReplicaX().Select(&postIds, "SELECT PostId FROM Reactions WHERE UserId = ?", userId)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get Reactions with userId=%s", userId)
+	}
+
+	txn, err := s.GetMasterX().Beginx()
 	if err != nil {
-		return
+		return err
 	}
-	deleted, err = result.RowsAffected()
-	return
+	defer finalizeTransactionX(txn, &err)
+
+	query := s.getQueryBuilder().
+		Delete("Reactions").
+		Where(sq.Eq{"UserId": userId})
```

Comment:
Unfortunately this query is very slow because it doesn't use the composite index, here's the explain on MySQL:
`EXPLAIN DELETE FROM Reactions WHERE UserId = 'ns3ggm5uipfz3kajwen9sxpbsr';`

|id |select_type|table|partitions|type       |possible_keys                      |key                                |key_len|ref|rows |filtered|Extra                                                             |
|---|-----------|-----|----------|-----------|-----------------------------------|-----------------------------------|-------|---|-----|--------|------------------------------------------------------------------|
|1  |DELETE     |Reactions|          |ALL        |                                   |                                   |       |   |2365792|100.00  |Using where                                                       |

**Execution time:** 28.918 s. 248 rows affected

I could change this to batch delete in 1000s but also use a SELECT/DELETE combination in order to better utilize the composite index:
1. `EXPLAIN SELECT PostId FROM Reactions WHERE UserId = 'kwrxn7sb4tfkbm9nnkbzshsmdy' LIMIT 1000;
`

|id |select_type|table|partitions|type       |possible_keys                      |key                                |key_len|ref|rows |filtered|Extra                                                             |
|---|-----------|-----|----------|-----------|-----------------------------------|-----------------------------------|-------|---|-----|--------|------------------------------------------------------------------|
|1  |SIMPLE     |Reactions|          |index      |PRIMARY,idx_reactions_channel_id   |idx_reactions_channel_id           |106    |   |2365543|10.00   |Using where; Using index                                          |

**Execution time:** 1.565 s. 253 rows returned

2. `EXPLAIN DELETE FROM Reactions WHERE PostId IN ('5afkp1w5t3nb3qkkby9g5gahsa', 'ob6wps1d8ibw9nhupj78436e8h') AND UserId = 'kwrxn7sb4tfkbm9nnkbzshsmdy';`

|id |select_type|table|partitions|type       |possible_keys                      |key                                |key_len|ref        |rows |filtered|Extra                                                             |
|---|-----------|-----|----------|-----------|-----------------------------------|-----------------------------------|-------|-----------|-----|--------|------------------------------------------------------------------|
|1  |DELETE     |Reactions|          |range      |PRIMARY                            |PRIMARY                            |212    |const,const|2    |100.00  |Using where                                                       |

**Execution time:** 9ms. 2 rows affected

Reply:
Change made, also added a sleep